pyteal/ast/abi/util_test.py

```diff
@@ -772,3 +772,250 @@ def test_sdk_type_specs_from_signature(sig_str, sig_args, sig_rets):
+def test_type_spec_is_assignable_safe_bidirectional_full_coverage(ts: type):
+    def exists_in_safe_bidirectional(_ts: type):
+        for safe_bidirectional in SAFE_BIDIRECTIONAL_TEST_CASES:
+            for t in safe_bidirectional.xs:
+                if type(t) == _ts:
+                    return True
+        return False
+
+    if isabstract(ts):
+        return
+    assert exists_in_safe_bidirectional(ts)
+
+
+class SafeAssignment(NamedTuple):
+    a: abi.TypeSpec
+    bs: list[abi.TypeSpec]
+
+
+SAFE_ASSIGNMENT_TEST_CASES: list[SafeAssignment] = [
+    SafeAssignment(
+        abi.StringTypeSpec(),
+        [abi.DynamicBytesTypeSpec(), abi.DynamicArrayTypeSpec(abi.ByteTypeSpec())],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(NamedTDecl),
+        [
+            abi.type_spec_from_annotation(
+                abi.Tuple3[
+                    abi.Uint64,
+                    abi.Tuple3[
+                        abi.Transaction,
+                        abi.Address,
+                        abi.StaticArray[abi.Byte, Literal[16]],
+                    ],
+                    abi.Transaction,
+                ]
+            ),
+        ],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(
+            abi.Tuple3[
+                abi.Uint64,
+                abi.Tuple3[
+                    abi.PaymentTransaction,
+                    abi.Address,
+                    abi.StaticArray[abi.Byte, Literal[16]],
+                ],
+                abi.PaymentTransaction,
+            ]
+        ),
+        [abi.type_spec_from_annotation(NamedTDecl)],
+    ),
+    SafeAssignment(
+        abi.AddressTypeSpec(),
+        [abi.StaticArrayTypeSpec(abi.ByteTypeSpec(), 32), abi.StaticBytesTypeSpec(32)],
+    ),
+    SafeAssignment(
+        abi.type_spec_from_annotation(NamedTComp0),
+        [abi.type_spec_from_annotation(NamedTComp1)],
+    ),
+] + [
+    SafeAssignment(spec, [abi.TransactionTypeSpec()])
+    for spec in abi.TransactionTypeSpecs
+    if spec != abi.TransactionTypeSpec()
+]
+
+
+@pytest.mark.parametrize("tc", SAFE_ASSIGNMENT_TEST_CASES)
+def test_type_spec_is_assignable_safe_assignment(tc: SafeAssignment):
+    assert len(tc.bs) > 0
+    for b in tc.bs:
+        assert abi.type_spec_is_assignable_to(tc.a, b)
+        assert not abi.type_spec_is_assignable_to(b, tc.a)
+
+
+@pytest.mark.parametrize("ts", bfs_on_inheritance(abi.TypeSpec))
+def test_type_spec_is_assignable_safe_assignment_full_coverage(ts: type):
+    def exists_in_safe_assignment(_ts: type):
+        for safe_assignment in SAFE_ASSIGNMENT_TEST_CASES:
+            if type(safe_assignment.a) == _ts:
+                return True
+            for t in safe_assignment.bs:
+                if type(t) == _ts:
+                    return True
+        return False
+
+    if isabstract(ts) or all(isabstract(parent) for parent in ts.__bases__):
+        return
+    assert exists_in_safe_assignment(ts)
```

Comment:
@ahangsu Optional:  What do you think of the following change?

* When reviewing the highlighted logic, I became a little worried that we may later add a new type and it'll _unintentionally_ go through the `return` branch.
* The proposal is to enforce that _absence_ is a deliberate decision by manually maintaining a list of types allowed to be absent.
* If you feel confident the as is approach provides sufficient coverage, then feel welcomed to resolve as is.

```suggestion
    # Abstract types and types without safe assignments should _not_ appear in test cases.
    if isabstract(ts) or ts in [
        abi.AccountTypeSpec,
        abi.ApplicationTypeSpec,
        abi.AssetTypeSpec,
        abi.BoolTypeSpec,
        abi.ByteTypeSpec,
        abi.Uint8TypeSpec,
        abi.Uint16TypeSpec,
        abi.Uint32TypeSpec,
        abi.Uint64TypeSpec,
    ]:
        assert not exists_in_safe_assignment(ts)
    else:
        assert exists_in_safe_assignment(ts)
```

Reply:
I can take this change, and I will try to explain more on comment on the choice of the set of type specs to skip.